src/picongpu/include/plugins/particleDetector/ParticleDetector.hpp

```diff
@@ -0,0 +1,542 @@
+/**
+ * Copyright 2016 Heiko Burau
+ *
+ * This file is part of PIConGPU.
+ *
+ * PIConGPU is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PIConGPU is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "ParticleDetectorFunctors.hpp"
+#include "ParticleDetector.kernel"
+
+#include "traits/PICToSplash.hpp"
+#include "plugins/ISimulationPlugin.hpp"
+#include "cuSTL/container/DeviceBuffer.hpp"
+#include "cuSTL/container/HostBuffer.hpp"
+#include "cuSTL/algorithm/kernel/Foreach.hpp"
+#include "cuSTL/cursor/MultiIndexCursor.hpp"
+#include "cuSTL/algorithm/mpi/Reduce.hpp"
+#include "cuSTL/algorithm/host/Foreach.hpp"
+#include "particles/policies/ExchangeParticles.hpp"
+#include "math/Vector.hpp"
+#include "algorithms/math.hpp"
+#include <boost/shared_ptr.hpp>
+
+/* libSplash data output */
+#include <splash/splash.h>
+#include <boost/filesystem.hpp>
+#include <string>
+#include <iostream>
+#include <fstream>
+#include <stdlib.h>
+
+namespace picongpu
+{
+using namespace PMacc;
+
+namespace po = boost::program_options;
+
+
+/** Virtual particle detector plugin.
+ *
+ * (virtually) propagates and collects particles to infinite distance.
+ *
+ */
+template<class ParticlesType>
+class ParticleDetector : public ISimulationPlugin
+{
+private:
+    std::string name;
+    std::string prefix;
+    std::string foldername;
+    uint32_t notifyPeriod;
+    MappingDesc* cellDescription;
+    std::ofstream outFile;
+
+    uint32_t numBinsTheta;
+    uint32_t numBinsPhi;
+    uint32_t numBinsEnergy;
+    float_X minEnergy;
+    float_X maxEnergy;
+    bool logScale;
+    float_X openingTheta;
+    float_X openingPhi;
+    float_X maxTheta;
+    float_X maxPhi;
```

Comment:
You are using a fixed relation between the two quantities here `maxPhi = 0.5 * openingPhi`.
Is it thus necessary to store both as class member variables?

Would it be better to just store one of these variables as member and recalculate the other on a member function scope?
e.g.
```
template<class AngleType> 
AngleType getMaxAngle(AngleType openingAngle)
{
    return openingAngle * AngleType(0.5);
}

void someMemeberfunction()
{
    ...
    const float_X maxPhi = getMaxAngle(openingPhi);
    const float_X maxTheta = getMaxAngle(openingTheta);
    ...
}


@psychocoderHPC what do you think as a future maintainer of this code?